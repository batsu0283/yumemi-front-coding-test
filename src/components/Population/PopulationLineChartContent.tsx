import {
  CartesianGrid,
  Legend,
  Line,
  LineChart,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from "recharts";
import type { Populations } from "../../hooks/usePopulations";
import {
  type PopulationCategoriesValue,
  type PrefCode,
  prefectureColors,
} from "../../utils/constants";
import type { Prefecture } from "../../utils/resasApi";

type PopulationLineChartContentProps = {
  populations: Populations;
  selectedPopulationCategory: PopulationCategoriesValue;
  checkedPrefectures: Set<PrefCode>;
  prefectures: Prefecture[];
};

export const PopulationLineChartContent = ({
  populations,
  selectedPopulationCategory,
  checkedPrefectures,
  prefectures,
}: PopulationLineChartContentProps) => (
  <ResponsiveContainer
    width="100%"
    height={500}
    data-testid="recharts-container"
  >
    <LineChart
      data={populations[selectedPopulationCategory]}
      margin={{ top: 5, right: 20, left: 30, bottom: 5 }}
    >
      <CartesianGrid strokeDasharray="3 3" />
      <XAxis dataKey="year" />
      <YAxis />
      <Tooltip />
      {Array.from(checkedPrefectures).map((prefCode) => (
        <Line
          key={prefCode}
          type="linear"
          dataKey={prefCode}
          stroke={prefectureColors[prefCode]}
          name={
            prefectures.find((pref) => pref.prefCode === prefCode)?.prefName
          }
          unit="äºº"
        />
      ))}
      <Legend />
    </LineChart>
  </ResponsiveContainer>
);
