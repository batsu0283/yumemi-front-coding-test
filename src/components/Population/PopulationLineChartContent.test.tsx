import { render } from "@testing-library/react";
import type { Populations } from "../../hooks/usePopulations";
import type {
  PopulationCategoriesValue,
  PrefCode,
} from "../../utils/constants";
import type { Prefecture } from "../../utils/resasApi";
import { PopulationLineChartContent } from "./PopulationLineChartContent";

vi.mock("recharts", async () => {
  const actual = await vi.importActual<typeof import("recharts")>("recharts");
  return {
    ...actual,
    ResponsiveContainer: ({ children }: { children: React.ReactElement }) => (
      <actual.ResponsiveContainer width={500} height={500}>
        {children}
      </actual.ResponsiveContainer>
    ),
  };
});
describe("PopulationLineChartContent", () => {
  const mockPopulations: Populations = {
    total: [
      { year: 2020, 1: 1000, 2: 1500 },
      { year: 2021, 1: 1100, 2: 1600 },
    ],
    young: [
      { year: 2020, 1: 200, 2: 300 },
      { year: 2021, 1: 210, 2: 310 },
    ],
    workingAge: [
      { year: 2020, 1: 600, 2: 900 },
      { year: 2021, 1: 610, 2: 910 },
    ],
    elderly: [
      { year: 2020, 1: 200, 2: 300 },
      { year: 2021, 1: 210, 2: 310 },
    ],
  };

  const mockPrefectures: Prefecture[] = [
    { prefCode: 1, prefName: "北海道" },
    { prefCode: 2, prefName: "青森県" },
  ];

  const mockCheckedPrefectures = new Set<PrefCode>([1, 2]);
  const selectedPopulationCategory: PopulationCategoriesValue = "total";
  afterEach(() => {
    vi.restoreAllMocks();
  });

  beforeAll(() => {
    globalThis.ResizeObserver = class ResizeObserver {
      observe() {
        console.log("observe");
      }
      unobserve() {
        console.log("unobserve");
      }
      disconnect() {
        console.log("disconnect");
      }
    };
  });
  it("正常にレンダリングされること", () => {
    const { container } = render(
      <PopulationLineChartContent
        populations={mockPopulations}
        selectedPopulationCategory={selectedPopulationCategory}
        checkedPrefectures={mockCheckedPrefectures}
        prefectures={mockPrefectures}
      />,
    );
    const rechartsContainer = container.querySelector(
      ".recharts-responsive-container",
    );
    expect(rechartsContainer).toBeInTheDocument();
  });

  it("XAxis、YAxis、Legendがレンダリングされること", () => {
    const { container } = render(
      <PopulationLineChartContent
        populations={mockPopulations}
        selectedPopulationCategory={selectedPopulationCategory}
        checkedPrefectures={mockCheckedPrefectures}
        prefectures={mockPrefectures}
      />,
    );
    expect(
      container.querySelector(".recharts-cartesian-axis.xAxis"),
    ).toBeInTheDocument();
    expect(
      container.querySelector(".recharts-cartesian-axis.yAxis"),
    ).toBeInTheDocument();
    expect(
      container.querySelector(".recharts-legend-wrapper"),
    ).toBeInTheDocument();
  });

  it("Lineの数がcheckedPrefecturesの個数と合うこと", () => {
    const { container } = render(
      <PopulationLineChartContent
        populations={mockPopulations}
        selectedPopulationCategory={selectedPopulationCategory}
        checkedPrefectures={mockCheckedPrefectures}
        prefectures={mockPrefectures}
      />,
    );
    const lines = container.querySelectorAll(".recharts-line");
    expect(lines).toHaveLength(mockCheckedPrefectures.size);
  });

  it("Snapshot", () => {
    const { container } = render(
      <PopulationLineChartContent
        populations={mockPopulations}
        selectedPopulationCategory={selectedPopulationCategory}
        checkedPrefectures={mockCheckedPrefectures}
        prefectures={mockPrefectures}
      />,
    );
    // innerHTMLを使ってランダムな属性を除外
    expect(container.innerHTML).toMatchSnapshot();
  });
});
