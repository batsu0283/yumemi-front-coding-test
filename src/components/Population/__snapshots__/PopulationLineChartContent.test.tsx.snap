// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopulationLineChartContent > Snapshot 1`] = `
"<div class="recharts-responsive-container" style="width: 500px; height: 500px; min-width: 0;"><div class="recharts-wrapper" style="position: relative; cursor: default; width: 100%; height: 100%; max-height: 500px; max-width: 500px;"><svg class="recharts-surface" width="500" height="500" style="width: 100%; height: 100%;" viewBox="0 0 500 500"><title></title><desc></desc><defs><clipPath id="recharts16-clip"><rect x="90" y="5" height="460" width="390"></rect></clipPath></defs><g class="recharts-cartesian-grid"><g class="recharts-cartesian-grid-horizontal"><line stroke-dasharray="3 3" stroke="#ccc" fill="none" x="90" y="5" width="390" height="460" x1="90" y1="465" x2="480" y2="465"></line><line stroke-dasharray="3 3" stroke="#ccc" fill="none" x="90" y="5" width="390" height="460" x1="90" y1="350" x2="480" y2="350"></line><line stroke-dasharray="3 3" stroke="#ccc" fill="none" x="90" y="5" width="390" height="460" x1="90" y1="235" x2="480" y2="235"></line><line stroke-dasharray="3 3" stroke="#ccc" fill="none" x="90" y="5" width="390" height="460" x1="90" y1="120" x2="480" y2="120"></line><line stroke-dasharray="3 3" stroke="#ccc" fill="none" x="90" y="5" width="390" height="460" x1="90" y1="5" x2="480" y2="5"></line></g><g class="recharts-cartesian-grid-vertical"><line stroke-dasharray="3 3" stroke="#ccc" fill="none" x="90" y="5" width="390" height="460" x1="90" y1="5" x2="90" y2="465"></line><line stroke-dasharray="3 3" stroke="#ccc" fill="none" x="90" y="5" width="390" height="460" x1="480" y1="5" x2="480" y2="465"></line></g></g><g class="recharts-layer recharts-cartesian-axis recharts-xAxis xAxis"><line orientation="bottom" width="390" height="30" x="90" y="465" class="recharts-cartesian-axis-line" stroke="#666" fill="none" x1="90" y1="465" x2="480" y2="465"></line><g class="recharts-cartesian-axis-ticks"><g class="recharts-layer recharts-cartesian-axis-tick"><line orientation="bottom" width="390" height="30" x="90" y="465" class="recharts-cartesian-axis-tick-line" stroke="#666" fill="none" x1="90" y1="471" x2="90" y2="465"></line><text orientation="bottom" width="390" height="30" stroke="none" x="90" y="473" class="recharts-text recharts-cartesian-axis-tick-value" text-anchor="middle" fill="#666"><tspan x="90" dy="0.71em">2020</tspan></text></g><g class="recharts-layer recharts-cartesian-axis-tick"><line orientation="bottom" width="390" height="30" x="90" y="465" class="recharts-cartesian-axis-tick-line" stroke="#666" fill="none" x1="480" y1="471" x2="480" y2="465"></line><text orientation="bottom" width="390" height="30" stroke="none" x="480" y="473" class="recharts-text recharts-cartesian-axis-tick-value" text-anchor="middle" fill="#666"><tspan x="480" dy="0.71em">2021</tspan></text></g></g></g><g class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"><line orientation="left" width="60" height="460" x="30" y="5" class="recharts-cartesian-axis-line" stroke="#666" fill="none" x1="90" y1="5" x2="90" y2="465"></line><g class="recharts-cartesian-axis-ticks"><g class="recharts-layer recharts-cartesian-axis-tick"><line orientation="left" width="60" height="460" x="30" y="5" class="recharts-cartesian-axis-tick-line" stroke="#666" fill="none" x1="84" y1="465" x2="90" y2="465"></line><text orientation="left" width="60" height="460" stroke="none" x="82" y="465" class="recharts-text recharts-cartesian-axis-tick-value" text-anchor="end" fill="#666"><tspan x="82" dy="0.355em">0</tspan></text></g><g class="recharts-layer recharts-cartesian-axis-tick"><line orientation="left" width="60" height="460" x="30" y="5" class="recharts-cartesian-axis-tick-line" stroke="#666" fill="none" x1="84" y1="350" x2="90" y2="350"></line><text orientation="left" width="60" height="460" stroke="none" x="82" y="350" class="recharts-text recharts-cartesian-axis-tick-value" text-anchor="end" fill="#666"><tspan x="82" dy="0.355em">400</tspan></text></g><g class="recharts-layer recharts-cartesian-axis-tick"><line orientation="left" width="60" height="460" x="30" y="5" class="recharts-cartesian-axis-tick-line" stroke="#666" fill="none" x1="84" y1="235" x2="90" y2="235"></line><text orientation="left" width="60" height="460" stroke="none" x="82" y="235" class="recharts-text recharts-cartesian-axis-tick-value" text-anchor="end" fill="#666"><tspan x="82" dy="0.355em">800</tspan></text></g><g class="recharts-layer recharts-cartesian-axis-tick"><line orientation="left" width="60" height="460" x="30" y="5" class="recharts-cartesian-axis-tick-line" stroke="#666" fill="none" x1="84" y1="120" x2="90" y2="120"></line><text orientation="left" width="60" height="460" stroke="none" x="82" y="120" class="recharts-text recharts-cartesian-axis-tick-value" text-anchor="end" fill="#666"><tspan x="82" dy="0.355em">1200</tspan></text></g><g class="recharts-layer recharts-cartesian-axis-tick"><line orientation="left" width="60" height="460" x="30" y="5" class="recharts-cartesian-axis-tick-line" stroke="#666" fill="none" x1="84" y1="5" x2="90" y2="5"></line><text orientation="left" width="60" height="460" stroke="none" x="82" y="5" class="recharts-text recharts-cartesian-axis-tick-value" text-anchor="end" fill="#666"><tspan x="82" dy="0.355em">1600</tspan></text></g></g></g><g class="recharts-layer recharts-line"><path stroke="#FF6633" name="北海道" stroke-width="1" fill="none" width="390" height="460" class="recharts-curve recharts-line-curve" stroke-dasharray="0px 0px" d="M90,177.5L480,148.75"></path></g><g class="recharts-layer recharts-line"><path stroke="#FFB399" name="青森県" stroke-width="1" fill="none" width="390" height="460" class="recharts-curve recharts-line-curve" stroke-dasharray="0px 0px" d="M90,33.75L480,5"></path></g></svg><div class="recharts-legend-wrapper" style="position: absolute; width: 450px; height: auto; left: 30px; bottom: 5px;"><ul class="recharts-default-legend" style="padding: 0px; margin: 0px; text-align: center;"><li class="recharts-legend-item legend-item-0" style="display: inline-block; margin-right: 10px;"><svg class="recharts-surface" width="14" height="14" style="display: inline-block; vertical-align: middle; margin-right: 4px;" viewBox="0 0 32 32"><title></title><desc></desc><path stroke-width="4" fill="none" stroke="#FF6633" d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16" class="recharts-legend-icon"></path></svg><span class="recharts-legend-item-text" style="color: rgb(255, 102, 51);">北海道</span></li><li class="recharts-legend-item legend-item-1" style="display: inline-block; margin-right: 10px;"><svg class="recharts-surface" width="14" height="14" style="display: inline-block; vertical-align: middle; margin-right: 4px;" viewBox="0 0 32 32"><title></title><desc></desc><path stroke-width="4" fill="none" stroke="#FFB399" d="M0,16h10.666666666666666
            A5.333333333333333,5.333333333333333,0,1,1,21.333333333333332,16
            H32M21.333333333333332,16
            A5.333333333333333,5.333333333333333,0,1,1,10.666666666666666,16" class="recharts-legend-icon"></path></svg><span class="recharts-legend-item-text" style="color: rgb(255, 179, 153);">青森県</span></li></ul></div><div tabindex="-1" class="recharts-tooltip-wrapper" style="visibility: hidden; pointer-events: none; position: absolute; top: 0px; left: 0px;"><div class="recharts-default-tooltip" style="margin: 0px; padding: 10px; background-color: rgb(255, 255, 255); border: 1px solid #ccc; white-space: nowrap;"><p class="recharts-tooltip-label" style="margin: 0px;"></p></div></div></div></div>"
`;
